### PR TITLE
Enable sharing model and environment from workspace to registry.

### PR DESCRIPTION
**Pull request by banibrata**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py [hunk 1]
```diff
@@ -373,43 +378,62 @@ def _get_latest_version(self, name: str) -> Environment:
...
+
         #  Get workspace info to get workspace GUID
         workspace = self._service_client.workspaces.get(
             resource_group_name=self._resource_group_name, workspace_name=self._workspace_name
         )
         workspace_guid = workspace.workspace_id
         workspace_location = workspace.location
 
-        # Get environment asset ID
+        # Get model asset ID
```
Comment (Man-MSFT) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:406]: ?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py [hunk 2]
```diff
@@ -120,7 +121,7 @@ def create_or_update(
             if self._registry_name:
                 # Case of copy model to registry
                 if isinstance(model, WorkspaceAssetReference):
-                    # verify that model is not already in registry
+                    ## verify that model is not already in registry
```
Comment (Man-MSFT) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:124]: extra #

**Review by luigiw — CHANGES_REQUESTED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py [hunk 3]
```diff
@@ -373,43 +378,62 @@ def _get_latest_version(self, name: str) -> Environment:
...
+        self._set_registry_client(registry_name)
+        return self.create_or_update(environment_ref)
+
+    def _set_registry_client(self, registry_name: str) -> None:
+        """Sets the registry client for the environment operations.
+
+        :param registry_name: Name of the registry.
+        :type registry_name: str
+        """
+        _client, rg, sub = get_registry_client(self._service_client._config.credential, registry_name)
```
Comment (Man-MSFT) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:430]: we don't not use the client, so we should name variable `client`

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py [hunk 4]
```diff
@@ -142,13 +143,16 @@ def create_or_update(
                             error_category=ErrorCategory.USER_ERROR,
                         )
 
-                    model = model._to_rest_object()
+                    model_rest_obj = model._to_rest_object()
```
Comment (Man-MSFT) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:146]: Out of consistency, the other file named the variable `environment_rest`
Not picky about which naming we use

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py [hunk 5]
```diff
@@ -373,43 +378,62 @@ def _get_latest_version(self, name: str) -> Environment:
...
 
-    # pylint: disable=no-self-use
-    def _prepare_to_copy(
-        self, environment: Environment, name: Optional[str] = None, version: Optional[str] = None
-    ) -> WorkspaceAssetReference:
-
-        """Returns WorkspaceAssetReference
-        to copy a registered environment to registry given the asset id
+    @monitor_with_activity(logger, "Environment.Share", ActivityType.PUBLICAPI)
+    def share(self, name, version, share_with_name, share_with_version, registry_name) -> Environment:
```
Comment (luigiw) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:382]: Can you add a `*` before `share_with_name` to make it a kw only argument. That will give us a chance to update the parameter name later.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py [hunk 6]
```diff
@@ -436,33 +440,24 @@ def list(
...
-            self._model_versions_operation,
-            self._resource_group_name,
-            self._workspace_name,
-        )
-        return Model._from_rest_object(result)
-
-    # pylint: disable=no-self-use
-    def _prepare_to_copy(
-        self, model: Model, name: Optional[str] = None, version: Optional[str] = None
-    ) -> WorkspaceAssetReference:
```
Comment (Man-MSFT) [inline sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:455]: Are we allowed to just delete this?
Or do we need to mark as deprecated instead

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py [hunk 7]
```diff
@@ -373,43 +378,62 @@ def _get_latest_version(self, name: str) -> Environment:
...
+        return self.create_or_update(environment_ref)
+
+    def _set_registry_client(self, registry_name: str) -> None:
+        """Sets the registry client for the environment operations.
+
+        :param registry_name: Name of the registry.
+        :type registry_name: str
+        """
+        _client, rg, sub = get_registry_client(self._service_client._config.credential, registry_name)
+        self._operation_scope.registry_name = registry_name
```
[the remark this answers is not in the document]
Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:431]: Good question: This should not change the ml_client context, this is only local to this operation context

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py (hunk 3 above)
Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:430]: Not sure I understood the question, _client is used in #434, the naming convention is at par with self._service_client.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py [hunk 8]
```diff
@@ -373,43 +378,62 @@ def _get_latest_version(self, name: str) -> Environment:
         )
         return Environment._from_rest_object(result)
 
-    # pylint: disable=no-self-use
-    def _prepare_to_copy(
```
[the remark this answers is not in the document]
Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:377]: I don't think we should keep it, if someone still wants they can use the existing SDK, and PMs are anyway updating the DOCs..

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py (hunk 6 above)
Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:455]: I don't think we should keep it, if someone still wants they can use the existing SDK, and PMs are anyway updating the DOCs.., are we keeping different versins of DOCS?

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:431]: Ok I see your point , let me check..

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (luigiw) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py:377]: Okay, sounds good.

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py (hunk 3 above)
Comment (Man-MSFT) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_environment_operations.py]: I mean, why did we add `_` prefix, when we do assign it, but not for `rg`/`sub` out of consistency.
so
```suggestion
client, rg, sub = get_registry_client(...)
```

File: sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py (hunk 6 above)
Comment (Man-MSFT) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:455]: I'm all for deleting, but just wanted to make sure we don't do anything we're not allowed to. It's `_` prefix but not sure if we have any other Azure rule

Comment (banibrata) [reply sdk/ml/azure-ai-ml/azure/ai/ml/operations/_model_operations.py:455]: This should not affect any Azure rule per my understanding, we keep changing private methods...as long as this is not exposed publicly

**Review by luigiw — APPROVED** [review]

**Pull request merged**